message/validation/validator.go

```diff
@@ -0,0 +1,467 @@
+	}
+
+	signerState.LastDecidedQuorumSize = len(signedMsg.Signers)
+
+	if err := signerState.MessageCounts.Validate(msg); err != nil {
+		return err
+	}
+
+	signerState.MessageCounts.Record(msg)
+
+	// Validate message counts within the current round.
+	if signerState.MessageCounts.Exceeds(maxMessageCounts(len(signedMsg.Signers))) {
+		return errors.New("too many messages")
+	}
+
+	return nil
+}
+
+func (mv *MessageValidator) maxRound(role spectypes.BeaconRole) specqbft.Round {
+	switch role {
+	case spectypes.BNRoleAttester, spectypes.BNRoleAggregator:
+		return 12 // TODO: consider calculating based on quick timeout and slow timeout
+	case spectypes.BNRoleProposer, spectypes.BNRoleSyncCommittee, spectypes.BNRoleSyncCommitteeContribution:
+		return 6
+	case spectypes.BNRoleValidatorRegistration:
+		return 0
+	default:
+		panic("unknown role")
+	}
+}
+
+func (mv *MessageValidator) currentEstimatedRound(role spectypes.BeaconRole, slot phase0.Slot) specqbft.Round {
+	slotStartTime := mv.netCfg.Beacon.GetSlotStartTime(slot)
+
+	firstRoundStart := slotStartTime.Add(mv.waitAfterSlotStart(role))
+	messageTime := time.Now() // TODO: attach it to the message when it's received
+
+	sinceFirstRound := messageTime.Sub(firstRoundStart)
+	if currentQuickRound := 1 + specqbft.Round(sinceFirstRound/roundtimer.QuickTimeout); currentQuickRound <= roundtimer.QuickTimeoutThreshold {
+		return currentQuickRound
+	}
+
+	sinceFirstSlowRound := messageTime.Sub(firstRoundStart.Add(time.Duration(roundtimer.QuickTimeoutThreshold) * roundtimer.QuickTimeout))
+	return roundtimer.QuickTimeoutThreshold + 1 + specqbft.Round(sinceFirstSlowRound/roundtimer.SlowTimeout)
+}
+
+func (mv *MessageValidator) waitAfterSlotStart(role spectypes.BeaconRole) time.Duration {
+	switch role {
+	case spectypes.BNRoleAttester, spectypes.BNRoleSyncCommittee:
+		return mv.netCfg.Beacon.SlotDurationSec() / 3
+	case spectypes.BNRoleAggregator, spectypes.BNRoleSyncCommitteeContribution:
+		return mv.netCfg.Beacon.SlotDurationSec() / 3 * 2
+	case spectypes.BNRoleProposer, spectypes.BNRoleValidatorRegistration:
+		return 0
+	default:
+		panic("unknown role")
+	}
+}
+
+func (mv *MessageValidator) validSigners(share *ssvtypes.SSVShare, msg *queue.DecodedSSVMessage) error {
+	// TODO: consider having different validation functions for each message type
+	if !share.BeaconMetadata.IsAttesting() {
+		return fmt.Errorf("validator not active")
+	}
+
+	contains := func(signer spectypes.OperatorID) func(operator *spectypes.Operator) bool {
+		return func(operator *spectypes.Operator) bool {
+			return operator.OperatorID == signer
+		}
+	}
+
+	switch m := msg.Body.(type) {
+	case *specqbft.SignedMessage:
+		if len(m.Signers) == 0 {
+			return ErrNoSigners
+		}
+
+		if len(m.Signers) == 1 {
+			qbftState := &specqbft.State{
+				Height: m.Message.Height,
+				Share:  &share.Share,
+			}
+			leader := specqbft.RoundRobinProposer(qbftState, m.Message.Round)
```

Comment:
I think this works but, even though there's a RoundRobinProposer in spec, in [proposal.go](https://github.com/bloxapp/ssv/blob/main/protocol/v2/qbft/instance/proposal.go#L214) a _proposer_ function is defined. If it changes, we might be in trouble. Would be better, if possible, to avoid this duplicated code. I don't know how. What do you think?

Reply:
`proposer` depends on `specqbft.State` and it's initialized in the validator controller, so I think it's too complex to use

Reply:
Indeed it is.